src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -0,0 +1,156 @@
+    const auto& parent = parent_and_leaf_input_side.first;
+    const auto& leaf_input_side = parent_and_leaf_input_side.second;
+    const auto leaf = parent->input(leaf_input_side);
+
+    // Collect all columns from the leaf that are actually referenced
+    auto referenced_leaf_columns = std::vector<std::shared_ptr<AbstractExpression>>{};
+    for (const auto& expression : leaf->column_expressions()) {
+      if (referenced_columns.find(expression) != referenced_columns.end()) {
+        referenced_leaf_columns.emplace_back(expression);
+      }
+    }
+
+    if (leaf->column_expressions().size() == referenced_leaf_columns.size()) continue;
+
+    // We cannot have a ProjectionNode that outputs no columns, so let's avoid that
+    if (referenced_leaf_columns.empty()) continue;
+
+    // If a leaf outputs columns that are never used, prune those columns by inserting a ProjectionNode that only
+    // contains the used columns
+    lqp_insert_node(parent, leaf_input_side, ProjectionNode::make(referenced_leaf_columns));
+    lqp_changed = true;
+  }
+
+  return lqp_changed;
+}
+
+void ColumnPruningRule::_search_for_projections_and_prune_columns(const std::shared_ptr<AbstractLQPNode>& lqp,
+                                                                  const ExpressionUnorderedSet& referenced_columns) {
+  /**
+   * Prune otherwise unused columns that are forwarded by ProjectionNodes
+   */
+
+  // First collect all the ProjectionNodes. Don't prune while visiting because visit_lqp() can't deal with nodes being
+  // replaced
+  auto projection_nodes = std::vector<std::shared_ptr<ProjectionNode>>{};
+  visit_lqp(lqp, [&](auto& node) {
+    if (node->type != LQPNodeType::Projection) return LQPVisitation::VisitInputs;
+    projection_nodes.emplace_back(std::static_pointer_cast<ProjectionNode>(node));
+    return LQPVisitation::VisitInputs;
+  });
+
+  // Replace ProjectionNodes with pruned ProjectionNodes if necessary
+  for (const auto& projection_node : projection_nodes) {
+    auto referenced_projection_expressions = std::vector<std::shared_ptr<AbstractExpression>>{};
+    for (const auto& expression : projection_node->node_expressions()) {
+      // We keep all non-column expressions
+      if (expression->type != ExpressionType::LQPColumn) {
+        referenced_projection_expressions.emplace_back(expression);
+      } else if (referenced_columns.find(expression) != referenced_columns.end()) {
+        referenced_projection_expressions.emplace_back(expression);
+      }
+    }
+
+    if (projection_node->node_expressions().size() == referenced_projection_expressions.size()) {
+      // No columns to prune
+      continue;
+    }
+
+    // We cannot have a ProjectionNode that outputs no columns
+    if (referenced_projection_expressions.empty()) {
+      lqp_remove_node(projection_node);
+    } else {
+      lqp_replace_node(projection_node, ProjectionNode::make(referenced_projection_expressions));
```

Comment:
Do we have to create a new projection node or could we also remove the columns from the existing one? Keeping the node would make it easier to follow the transformations across optimization steps.

Reply:
replacing expressions now, instead of inserting new node.